02-peer-protocol.md

```diff
@@ -430,6 +446,515 @@ would create a Denial of Service risk; therefore, forgetting it is recommended
+the `funding_txid`, instead of the `temporary_node_id`.
+
+
+### The `funding_signed2` Message
+
+This message contains witness data for for the inputs that were
+originally exchanged in `funding_add_input`.
+
+1. type: 64 (`funding_signed2`)
+2. data:
+    * [`channel_id`:`channel_id`]
+    * [`u16`:`num_witnesses`]
+    * [`num_witnesses*witness_stack`:`witness_stack`]
+
+1. subtype: `witness_stack`
+2. data:
+    * [`sha256`:`prevtx_txid`]
+    * [`u32`:`prevtx_vout`]
+    * [`u16`:`num_input_witness`]
+    * [`num_input_witness*witness_element`:`witness_element`]
+
+1. subtype: `witness_element`
+2. data:
+    * [`u16`:`len`]
+    * [`len*byte`:`witness`]
+
+#### Requirements
+The sending node:
+  - MUST verify it has received valid commitment signatures from its peer
+  - MUST set `witness` to the serialized witness data for the input
+    corresponding to `prevtx_txid`:`prevtx_vout`
+  - MUST remember the details of this funding transaction.
+  - MUST NOT send a `witness_stack` whose length exceeds its
+    previously indicated `max_witness_len`.
+  - if it has NOT received a valid `commitment_signed` message:
+    - MUST NOT send a `funding_signed2` message.
+
+The receiving node:
+  - if the `witness_stack` length exceeds `max_witness_len`:
+    - MUST error.
+  - MUST verify it has received valid commitment signatures
+  - SHOULD apply `witness`es to the funding transaction and
+    broadcast it.
+
+#### Rationale
+
+Every node must transmit their complete set of witness data for
+the inputs included in the funding transaction.
+
+`prevtx_txid` and `prevtx_vout` are the identifier for the input
+which the witness data pertain to
+
+`witness` is the data for a witness element in a witness stack
+
+### Kicking Off Replace-By-Fee: `init_rbf`
+
+This message initiates the flow to create a replacement funding
+transaction.  It is sent by the channel opener, after the funding
+transaction has been broadcast but before a `funding_locked` message
+is exchanged.
+
+Once an `init_rbf` message has been successfully ack'd by the accepter
+node, the message flow returns to `commitment_signed` and proceeds as
+indicated above.
+
+1. type: 62 (`init_rbf`)
+2. data:
+    * [`channel_id`:`channel_id`]
+    * [`u64`:`funding_satoshis`]
+    * [`u32`:`feerate_per_kw`]
+    * [`u32`:`feerate_per_kw_funding`]
+    * [`u16`:`num_additional_inputs`]
+    * [`num_additional_inputs*input_info`:`input_info`]
+    * [`u16`:`num_outputs`]
+    * [`num_outputs*output_info`:`output_info`]
+
+#### Requirements
+
+The sending node:
+  - MUST have sent `open_channel2`.
+  - MUST send a `feerate_per_kw_funding` greater than the most recently
+    negotiated rate.
+  - MAY update the `feerate_per_kw`, the commitment transaction feerate.
+  - MAY include additional inputs.
+  - MAY set the `num_inputs` to zero.
+  - MUST transmit all outputs, excluduing the channel funding output.
+  - MAY include at least one output with the `sats` set to zero,
```

Comment:
Should we also specify which output to reduce ?

Reply:
Well, retransmitting all of the outputs would give you the opportunity to update the outputs as necessary.

The change output is calculated dynamically by each of the peers, and is identifiable in the output set as "the output of value zero", so you wouldn't need to update it. In fact, the change calculation should 1) drop the change output if it's below dust and 2) add or subtract any resultant amount to the opener's satoshis in the funding output. So, you wouldn't need to update the outputs on an rbf, unless you wanted to preserve  your funding output's value.

Reply:
> So, you wouldn't need to update the outputs on an rbf, unless you wanted to preserve your funding output's value.

That's what I assumed: doesn't it make it more simple for both participants if the funding amount stayed the same but the fee bump was taken on the change output of the funder instead ?
Since both add inputs to the tx, they will most likely also include change outputs, at least that what I assumed when writing that.
My question was then whether to introduce an explicit MUST on the receiver side to check that the tx the funder sends does not greeds on the **fundee**'s change output (which implems would of course do, but ..)

But maybe I'm missing something here ?

Reply:
> That's what I assumed: doesn't it make it more simple for both participants if the funding amount stayed the same but the fee bump was taken on the change output of the funder instead ?

It's up to the funder to know what it's estimated change burden will be and add sufficient funds such that the total amount of funding is untouched by a fee-bump, if that's what they desire.

> does not greeds on the fundee's change output (which implems would of course do, but ..)

Fess fees never come out of the accepter's change; though it might (hypothetically) start eating out of their funding balance if the opener's balance is consumed by change. Reconsidering this and the point you made earlier, a `nack_rbf` message (or a 'soft-error') might be appropriate here. This would allow the second node to disallow a fee update that would eat into the entirety of the fundchannel amount.